matchbox_server/src/main.rs

```diff
@@ -32,39 +26,51 @@ async fn main() {
                 .with_target(false),
         )
         .init();
+}
+
+#[tokio::main]
+async fn main() {
+    // Initialize logger
+    setup_logging();
 
     // Parse clap arguments
     let args = Args::parse();
 
     // Setup router
-    let server_state = Arc::new(futures::lock::Mutex::new(ServerState::default()));
-    let app = Router::new()
-        .route("/health", get(health_handler))
-        .route("/", get(ws_handler))
-        .route("/:room_id", get(ws_handler))
-        .layer(
-            // Allow requests from anywhere - Not ideal for production!
-            CorsLayer::new()
-                .allow_origin(Any)
-                .allow_methods(Any)
-                .allow_headers(Any),
-        )
-        .layer(
-            // Middleware for logging from tower-http
-            TraceLayer::new_for_http().on_response(
-                DefaultOnResponse::new()
-                    .level(Level::INFO)
-                    .latency_unit(LatencyUnit::Micros),
-            ),
-        )
-        .with_state(server_state);
-
-    // Run server
     info!("Matchbox Signaling Server: {}", args.host);
-    axum::Server::bind(&args.host)
-        .serve(app.into_make_service_with_connect_info::<SocketAddr>())
+
+    let mut state = ServerState::default();
+    let server = SignalingServerBuilder::new(args.host, MatchmakingDemoTopology, state.clone())
+        .on_connection_request({
+            let mut state = state.clone();
+            move |connection| {
+                let room_id = RoomId(connection.path.clone().unwrap_or_default());
+                let next = connection
+                    .query_params
+                    .get("next")
+                    .and_then(|next| next.parse::<usize>().ok());
+                let room = RequestedRoom { id: room_id, next };
+                state.add_waiting_client(connection.origin, room);
+                Ok(true) // allow all clients
+            }
+        })
+        .on_id_assignment({
+            move |(origin, peer_id)| {
+                info!("Client connected {origin:?}: {peer_id:?}");
+                state.assign_id_to_waiting_client(origin, peer_id);
+            }
+        })
+        .cors()
+        .trace()
+        .mutate_router(|router| {
+            // Apply router transformations
+            router.route("/health", get(health_handler))
```

Comment:
Probably neater as a closure
```suggestion
            router.route("/health", get(|| async { StatusCode::OK }))
```